README.md
# 360EAVP

A 360-degree Edition-Aware Video Player (360EAVP), a web browser-based open-source application for tile-based DASH streaming and visualization of 360-degree edited videos on HMDs. The proposed application is built on top of the [VR DASH Tile-Based Player (VDTP)](https://github.com/ForgetMe17/vr-dash-tile-player).

This project is an extension of its previous version and leverages its usability implementing more functionalities, such as:

1. Integration with video editing. (Snap-change and Fade-rotation implemented)

2. Viewport prediction module. (Linear and Ridge Regression)

3. Center of the viewport track for HMDs.

4. New approach the retrieve visible faces and its visibility percentage to be used on ABR algorithms.
5. Data collection and log generation module during video playback.

  

## Tools and Frameworks we use

  

1. dash.js - For adaptive bitrate streaming via DASH.

3. aframe - For 3D rendering (WebXR supported for any videos based on VP9).

4. angular.js - For data virtualization and code optimization.
  

## ABR Algorithms we use


1. FOVEditRule - Choose the bitrate according to the FOV and Edit information.

2. FOVRule - Choose the bitrates for each tile according to FOV.

3. LowestBitrateRule - Always choose the lowest bitrate for each tile.

4. HighestBitrateRule - Always choose the highest bitrate for each tile.

5. DefaultRule - Using default ABR rules by dash.js (observing each tile's stats independently).

  

Default: Using FOVEditRule as default. If need please change the config in HTML page.

  

## How to run

1. Run the HTML file via HTTP address (e.g., http://localhost/360eavp/Index.html).

2. Confirm the location of JSON file, the Mode (VOD/LIVE) and the ABR Rule you want to use in HTML page, then click "link".

3. Click "Render" to create the 3D environment.

4. Click "Load" to initialize MediaPlayer according to JSON description.

5. Click "Play" and "Pause" to control the player.

6.After the playback end, a .CSV file with information per frame of the playback should be start to be downloaded. 

  

## Media Preprocessing

  

1. DASH Adaptive Streaming for HTML 5 Video (Webm) : https://developer.mozilla.org/en-US/docs/Web/Media/DASH_Adaptive_Streaming_for_HTML_5_Video

2. FFMPEG + Bento4 (MP4) : http://www.ffmpeg.org/documentation.html , https://www.bento4.com/documentation/

3. If you want to load the media files locally, you need to set up a HTTP file server on your device. [XAMPP](https://www.apachefriends.org/pt_br/index.html) is an option for an APACHE server.  
  

#### Here is an example of media preprocessing for 360EAVP platform with an ERP video. The file on `scripts/pre_processing.sh` provides a script for doing this pre-processing. (if you have a CMP video, just skip the steps of ERP/CMP converting.)

1. Make sure you have installed the tools we need to use: FFMPEG, Bento4.

2. Use FFMPEG to convert your video from ERP mode to CMP mode (if you have a CMP video, just skip this step). Here's the code for converting an ERP video to a CMP video with the size 4320 x 2880 (each face with 1440 x 1440) using VP9 coder (all the parameters can be customized according to your requirement, more info please check FFMPEG official documentation):

```

ffmpeg -i ERP_video.mp4 -vf v360=e:c3x2:cubic:w=4320:h=2880:out_pad=0 -c:v libvpx-vp9 -crf 0 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an CMP_video.mp4

```

3. The output video from step 2 should be a single video with 2 x 3 faces embedded together. Use FFMPEG to slice it into 6 video files:

```

ffmpeg -y -i CMP_video.mp4 -vf "crop=w=in_w/3:h=in_h/2:x=0*(in_w/3):y=0*(in_h/2)" -c:v libvpx-vp9 -keyint_min 30 -g 30 -sc_threshold 0 -an face0.mp4

ffmpeg -y -i CMP_video.mp4 -vf "crop=w=in_w/3:h=in_h/2:x=1*(in_w/3):y=0*(in_h/2)" -c:v libvpx-vp9 -keyint_min 30 -g 30 -sc_threshold 0 -an face1.mp4

ffmpeg -y -i CMP_video.mp4 -vf "crop=w=in_w/3:h=in_h/2:x=2*(in_w/3):y=0*(in_h/2)" -c:v libvpx-vp9 -keyint_min 30 -g 30 -sc_threshold 0 -an face2.mp4

ffmpeg -y -i CMP_video.mp4 -vf "crop=w=in_w/3:h=in_h/2:x=0*(in_w/3):y=1*(in_h/2)" -c:v libvpx-vp9 -keyint_min 30 -g 30 -sc_threshold 0 -an face3.mp4

ffmpeg -y -i CMP_video.mp4 -vf "crop=w=in_w/3:h=in_h/2:x=1*(in_w/3):y=1*(in_h/2)" -c:v libvpx-vp9 -keyint_min 30 -g 30 -sc_threshold 0 -an face4.mp4

ffmpeg -y -i CMP_video.mp4 -vf "crop=w=in_w/3:h=in_h/2:x=2*(in_w/3):y=1*(in_h/2)" -c:v libvpx-vp9 -keyint_min 30 -g 30 -sc_threshold 0 -an face5.mp4

```

4. The output videos from step 3 should be six videos as six faces in CMP mode. Use FFMPEG to trancode the videos in different bitrates. ```-crf``` is used to control the video's quality (the parameter smaller, the quality better):

```

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 60 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face0_60.mp4

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 40 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face0_40.mp4

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 20 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face0_20.mp4

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 0 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face0_0.mp4

...

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 60 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face5_60.mp4

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 40 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face5_40.mp4

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 20 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face5_20.mp4

ffmpeg -i face0.mp4 -c:v libvpx-vp9 -crf 0 -b:v 0 -keyint_min 30 -g 30 -sc_threshold 0 -an face5_0.mp4

```

5. The output videos from step 4 should be six sets of videos with different qualities. Use Bento4 to convert them into fragment format for segmentation. ```--fragment-duration``` is used to set the length of each segment (millisecond):

```

mp4fragment --fragment-duration 1000 face0_60.mp4 f_face0_60.mp4

mp4fragment --fragment-duration 1000 face0_40.mp4 f_face0_40.mp4

mp4fragment --fragment-duration 1000 face0_20.mp4 f_face0_20.mp4

mp4fragment --fragment-duration 1000 face0_0.mp4 f_face0_0.mp4

...

mp4fragment --fragment-duration 1000 face5_60.mp4 f_face5_60.mp4

mp4fragment --fragment-duration 1000 face5_40.mp4 f_face5_40.mp4

mp4fragment --fragment-duration 1000 face5_20.mp4 f_face5_20.mp4

mp4fragment --fragment-duration 1000 face5_0.mp4 f_face5_0.mp4

```

6. Use Bento4 to convert them into segments for DASH streaming:

```

mp4dash --output-dir=face0 --mpd-name=face0.mpd f_face0_60.mp4 f_face0_40.mp4 f_face0_20.mp4 f_face0_0.mp4

...

mp4dash --output-dir=face5 --mpd-name=face5.mpd f_face5_60.mp4 f_face5_40.mp4 f_face5_20.mp4 f_face5_0.mp4

```

7. If audio is necessary, please follow the steps to extract it from the original video file. Here we set audio as an independent track, you can also embed it into any one of faces' tracks:

```

ffmpeg -i ERP_video.mp4 -vn -acodec copy -y audio.mp4

mp4fragment --fragment-duration 1000 audio.mp4 f_audio.mp4

mp4dash --output-dir=audio --mpd-name=audio.mpd f_audio.mp4

```

8. Here we finish all the steps. A JSON file would be necessary if you want to play it more convenient in our platform. The file `default.json` brings the structure that needs to be followed. It is worth to mention that although the 360EAVP was design to work with edits, they are optional. If audio is unavailable, please change it as ```"audio": ""``` . If one wants to use edits on the video, please fill in the ```edit``` key with the ```frame``` that you want it to happpen, the ```type``` (```instant``` for Snap-change and ```gradual``` for Fade-rotation), and the normalized region of interest of the edit.

```
{
	"baseUrl": "[Your file location]",
	"face": 6,
	"row": 1,
	"col": 1,
	"duration": 1,
    "tiles": [
		[[{"src": "face0/face0.mpd", "width": "1440", "height": "1440", "x": 720, "y": 0, "z": 0, "rx": 0, "ry": 270, "rz": 0}]],
		[[{"src": "face1/face1.mpd", "width": "1440", "height": "1440", "x": -720, "y": 0, "z": 0, "rx": 0, "ry": 90, "rz": 0}]],
		[[{"src": "face2/face2.mpd", "width": "1440", "height": "1440", "x": 0, "y": 720, "z": 0, "rx": 90, "ry": 0, "rz": 180}]],
		[[{"src": "face3/face3.mpd", "width": "1440", "height": "1440", "x": 0, "y": -720, "z": 0, "rx": -90, "ry": 0, "rz": 180}]],
		[[{"src": "face4/face4.mpd", "width": "1440", "height": "1440", "x": 0, "y": 0, "z": -720, "rx": 0, "ry": 0, "rz": 0}]],
		[[{"src": "face5/face5.mpd", "width": "1440", "height": "1440", "x": 0, "y": 0, "z": 720, "rx": 0, "ry": 180, "rz": 0}]]
	],
	"audio": "audio/audio.mpd",
	"edits":
		{
		   "edit" : [
		      {
		         "frame" : 2080,
		         "type": "gradual",
		         "region_of_interest": [                                                                 
		            {
		            "rank": 1,
		            "ROI_theta": 0.267
		            }
		         ]     
		      }
		   ]
		}
}
```

  
(Preprocessed dataset Guitar_Man in Google Drive: https://drive.google.com/file/d/1T3NTSe8v4dRodUizYnejCgaaCoRmSMV9/view?usp=share_link)

Original video available at: https://youtu.be/_v0yuYHiqOY

( Preprocessed dataset CQ07 in Google Drive from VDTP: https://drive.google.com/file/d/1yQpNv4Mpe1OQzgXb6epJdvo44YGX3jDH/view?usp=sharing )



dash.js/changelog.txt
Base version {

    dash.js v4.3.0; 

}

Modifications {

    Add: [Settings.js] { 
        info: {
            id: null,
            count: null,
            face: null,
            row: null,
            col: null,
            duration: null,
            width: null,
            height: null,
            location: {
                'x': null,
                'y': null,
                'z': null
            },
            rotation: {
                'rx': null,
                'ry': null,
                'rz': null
            },
            totalThroughputNeeded: false
        },
    };
    Add: [AbrController.js] AbrController.getSettings();
    Add: [BufferController.js] BufferController.triggerEvent(...) { payload.info = settings.get().info; };
    Add: [ThroughputHistory.js] const AVERAGE_THROUGHPUT_SAMPLE_AMOUNT_TOTALTHROUGHPUTNEEDED = 1;
    Add: [ThroughputHistory.js] ThroughputHistory.push(...) { ...
                                    if (settings.get().info.totalThroughputNeeded && $scope !== undefined && $scope.requestList !== undefined) {
                                        $scope.requestList.push(httpRequest);
                                    }
                                ... };
    Add: [ThroughputHistory.js] var appElement = document.querySelector('[ng-controller=DashController]');
                                var $scope = window.angular.element(appElement).scope();
    Add: [PlaybackController.js] var appElement = document.querySelector('[ng-controller=DashController]');
                                 var $scope = window.angular.element(appElement).scope();                             

    Change: [PlaybackController.js] PlaybackController.onPlaybackProgression() {  // Consider the situation whatever isDynamic is
                                        if (
                                            //isDynamic &&
                                            _isCatchupEnabled() &&
                                            settings.get().streaming.liveCatchup.playbackRate > 0 &&
                                            !isPaused() &&
                                            !isSeeking()
                                        ) {
                                            if (_needToCatchUp()) {
                                                if ($scope !== undefined && $scope.playerCatchUp !== undefined) {
                                                    $scope.playerCatchUp[settings.get().count] = true;
                                                }
                                                startPlaybackCatchUp();
                                            } else {
                                                if ($scope !== undefined && $scope.playerCatchUp !== undefined) {
                                                    $scope.playerCatchUp[settings.get().count] = false;
                                                }
                                                stopPlaybackCatchUp();
                                            }
                                        }
                                    };
    Change: [PlaybackController.js] PlaybackController.getCurrentLiveLatency() {  // Change totally with compatibility for non-isDynamic
                                        if (isNaN(availabilityStartTime)) {
                                            return NaN;
                                        }
                                        let currentTime = getNormalizedTime();
                                        if (isNaN(currentTime) || currentTime === 0) {
                                            return 0;
                                        }

                                        if (!isDynamic && $scope !== undefined && $scope.normalizedTime !== undefined) {
                                            const now = $scope.normalizedTime * 1000 + timelineConverter.getClientTimeOffset() * 1000;
                                            return Math.max(((now - availabilityStartTime - currentTime * 1000) / 1000).toFixed(3), 0);
                                        }
                                        
                                        const now = new Date().getTime() + timelineConverter.getClientTimeOffset() * 1000;
                                        return Math.max(((now - availabilityStartTime - currentTime * 1000) / 1000).toFixed(3), 0);
                                    };
    Change: [ThroughputHistory.js] ThroughputHistory.getSampleSize(...) { ... if (isThroughput) { sampleSize = settings.get().info.totalThroughputNeeded ? AVERAGE_THROUGHPUT_SAMPLE_AMOUNT_TOTALTHROUGHPUTNEEDED : isLive ? AVERAGE_THROUGHPUT_SAMPLE_AMOUNT_LIVE : AVERAGE_THROUGHPUT_SAMPLE_AMOUNT_VOD; } ... };
    Change: [Settings.js]     function mixinSettings(source, dest, path) {
        for (let n in source) {
            if (source != null && source.hasOwnProperty(n)) {
                if (dest != null && dest.hasOwnProperty(n)) {
                    if (typeof source[n] === 'object' && source[n] !== null) {
                        mixinSettings(source[n], dest[n], path.slice() + n + '.');
                    } else {
                        dest[n] = Utils.clone(source[n]);
                    }
                } else {
                    console.error('Settings parameter ' + path + n + ' is not supported');
                }
            }
        }
    }
}

edit/dynamicEdit.js
// Dynamic Edit with SnapCut and Fade Rotation

const INSTANT_EDIT = "instant";
const GRADUAL_EDIT = "gradual";

const EDIT_TIME = 1.5; //time in seconds that the edit will take
const FADE_TIME = EDIT_TIME/2; //time in seconds after edit started to start the fade animation
const ROTATION_SPEED = 0.002; //radians per frame (60 frames per second on web)
const OPACITY_THRESHOLD = 0.95; // opacity that will start the snapcut on the fade rotation edit
const ENABLE_EDIT = true;



var sphere_reference, camera_reference;
var doIt = true;
var doFade = true;

var enableRotation = true;
var enableFade = true;
var fadeStarted = false;
var fadePaused = false;
var isSnapCutEnabled = true;
var dist_nearest_roi_gradual;
var direction;
var isRotating = false;
var editEditInfoJSON;
let next_edit = 0;
let CvpXRadians, CvpYRadians;
let last_frame_with_edit = 0;
    
function dynamicEditClass () {
        
    [sphere_reference, camera_reference ] = getIframeEntities('frame');
    
    var appElement = document.querySelector('[ng-controller=DashController]');
    var $scope = angular.element(appElement).scope();
    editEditInfoJSON = $scope.contents.edits && $scope.contents.edits.edit;
    
    if ( $scope.frameNumber == 0) {
        $scope.frameNumber = camera_reference.videoFrames["video_0"];
        $scope.videoFrameRate = camera_reference.videoFrames["video_0"]["frameRate"];
    }

    let isVideoRunning = $scope.buffer_empty_flag.filter(empty => empty == true).length > 0 ? false : true;

    if (ENABLE_EDIT && isVideoRunning && editEditInfoJSON){
    
        var currentFrame = $scope.frameNumber.get();
        var frameRate = $scope.videoFrameRate;

        let CvpRadians = sphere_reference["head_movement_radians"];
        
        if (CvpRadians != undefined){

            
            // Only use the X axis for the edit
            CvpXRadians = CvpRadians[0];
            $scope.yaw = CvpRadians[0];
            CvpYRadians = CvpRadians[1];
            $scope.pitch = CvpRadians[1];     
    
            let sphereOpacity = sphere_reference.object3DMap.mesh.material.opacity;

            //Fade Effect Showcase
            if (editEditInfoJSON[next_edit] && editEditInfoJSON[next_edit]["type"] == GRADUAL_EDIT){

                if (fadeStarted && fadePaused){
                    console.log("RESUME")
                    sphere_reference.emit('resumeFadeEffect');
                    fadePaused = false;
                }
                
                let currentFrameEdit = editEditInfoJSON[next_edit] ? editEditInfoJSON[next_edit]["frame"] : 0;
                let frameStartGradualRotation = Math.ceil(currentFrameEdit - (EDIT_TIME * 2 * frameRate));
                let frameStopGradualRotation = Math.ceil(currentFrameEdit + (EDIT_TIME * 0.5 * frameRate));
                let frameStartFade = Math.ceil(frameStopGradualRotation - ((EDIT_TIME + FADE_TIME)* 2 * frameRate));
                
                if (currentFrame >= frameStartGradualRotation  && enableRotation ){
                    console.log("START ROTATION");
                    
                    //Predict where the user will be looking on the Edit frame to know if the edit will happen or not

                    let predictedViewport = $scope.predict_center_viewport(currentFrameEdit - currentFrameEdit);
                    let predict_center_viewport_x = predictedViewport[0];;

                    [dist_nearest_roi_gradual, direction,] = getNearestRegionOfInterest(predict_center_viewport_x);
                    


                    // if the RoI is under 30° from the current center of the viewport, it is not needed to do an edit
                    if (dist_nearest_roi_gradual >= 0.5235 ){
                        isRotating = true;
                    }

                    enableRotation = false;
                }

    
                if (currentFrame >= frameStartFade && enableFade && isRotating){
                    console.log("START FADE");
                    sphere_reference.emit('startFadeEffect');
                    enableFade = false;
                    fadeStarted = true;
                }

                if (sphereOpacity >= OPACITY_THRESHOLD && fadeStarted && doIt){
                    console.log("START SNAPCUT");
                    doIt = false;
                    let snapCutRotation;
                    let delay;
                    if (direction > 0){
                        delay = -ROTATION_SPEED * (frameRate / 2);
                    }
                    else {
                        delay = ROTATION_SPEED * (frameRate / 2);
                    }
                    snapCutRotation = handleSnapCut(CvpXRadians - delay);

                    

                    if (snapCutRotation != undefined){
                        $scope.radiansRotationValue = snapCutRotation;
                        $scope.editHappenedValue = true;
                    }
                }
    
                if (isRotating){
                    $scope.editTypeValue = GRADUAL_EDIT;
                    $scope.hasEditScheduledValue = true;

                    const quaternion = new THREE.Quaternion();
                    

                    if (direction > 0)
                    {
                        quaternion.setFromAxisAngle( new THREE.Vector3( 0, 1, 0 ), ROTATION_SPEED );
                        camera_reference.object3D.quaternion.multiply(quaternion);
                        sphere_reference.object3D.quaternion.multiply(quaternion);

                        const quaternionFace = new THREE.Quaternion();
                        quaternionFace.setFromAxisAngle( new THREE.Vector3( 0, 1, 0 ), ROTATION_SPEED );
                
                        for (var face in camera_reference.faceStructure) {
                            for (var position = 0; position < camera_reference.faceStructure[face].length; position++){
                                camera_reference.faceStructure[face][position].applyQuaternion(quaternionFace);
                                
                            }
                        }
                    }
                    else {
                        quaternion.setFromAxisAngle( new THREE.Vector3( 0, 1, 0 ),-1*ROTATION_SPEED );
                        camera_reference.object3D.quaternion.multiply(quaternion);
                        sphere_reference.object3D.quaternion.multiply(quaternion);

                        const quaternionFace = new THREE.Quaternion();
                        quaternionFace.setFromAxisAngle( new THREE.Vector3( 0, 1, 0 ), -1*ROTATION_SPEED );
                
                        for (var face in camera_reference.faceStructure) {
                            for (var position = 0; position < camera_reference.faceStructure[face].length; position++){
                                camera_reference.faceStructure[face][position].applyQuaternion(quaternionFace);
                                
                            }
                        }
                    }
    
                }

                if (currentFrame >= frameStopGradualRotation){
                    console.log("STOP ROTATION");
                    next_edit++;
                    isRotating = false;
                    enableRotation = true;
                    enableSnapCut = true;
                    enableFade = true;
                    fadeStarted = false;
                    doIt = true;

                    $scope.editTypeValue = "null";
                    $scope.radiansRotationValue = 0;
                    $scope.hasEditScheduledValue = false;
                    $scope.editHappenedValue = false;
                }
            }
        
            
            // Check if there is another edit to be done, if the current frame has an edit on the ediInfo file 
            // and if the last frame already had an edit (this is needed because the last frame is redrawn when the edit happens)
            if (editEditInfoJSON[next_edit] && editEditInfoJSON[next_edit]["type"] == INSTANT_EDIT)
            {
                let currentFrameEdit = editEditInfoJSON[next_edit] ? editEditInfoJSON[next_edit]["frame"] : 0;
                
                if (currentFrame >= currentFrameEdit && last_frame_with_edit != currentFrame ){
                    console.log("START INSTANT SNAPCUT");
                    $scope.editTypeValue = INSTANT_EDIT;
                    $scope.hasEditScheduledValue = true;

                    let snapCutRotation = handleSnapCut(CvpXRadians);
                    last_frame_with_edit = currentFrame;
                    if (snapCutRotation != undefined){
                        $scope.radiansRotationValue = snapCutRotation;
                        $scope.editHappenedValue = true;
                    }
                    next_edit++;
                }
            }
        }
    }
    else if (fadeStarted && !isVideoRunning){
        console.log("PAUSE")
        fadePaused = true;
        sphere_reference.emit('pauseFadeEffect');
    }
    console.log("fadeStarted:", fadeStarted)
    requestAnimationFrame(dynamicEditClass);
}


function getIframeEntities(frameId) {
    var frameObj = document.getElementById(frameId);
    if (frameObj){
        var camera_reference = frameObj.contentWindow.document.querySelector('#video_camera');
        var sphere_reference = frameObj.contentWindow.document.querySelector('#sky-sphere'); 
        return [sphere_reference, camera_reference];
    }
    return;
    }


function fireRotation (roi_radians)
    {
        const quaternion = new THREE.Quaternion();
        quaternion.setFromAxisAngle( new THREE.Vector3( 0, 1, 0 ), roi_radians );
        console.log("fire Rotation!!");
        console.log("BEFORE camera_reference.faceStructure", camera_reference.faceStructure);
        const quaternionFace = new THREE.Quaternion();
        quaternionFace.setFromAxisAngle( new THREE.Vector3( 0, 1, 0 ), roi_radians );

        for (var face in camera_reference.faceStructure) {
            for (var position = 0; position < camera_reference.faceStructure[face].length; position++){
                camera_reference.faceStructure[face][position].applyQuaternion(quaternionFace);
                
            }
        }
        console.log("AFTER camera_reference.faceStructure", camera_reference.faceStructure);
        camera_reference.object3D.quaternion.multiply(quaternion);
        sphere_reference.object3D.quaternion.multiply(quaternion);
    }


function handleSnapCut(CvpXRadians)
{
    let abs_dist_nearest_roi, dist_nearest_roi, index_dist_nearest_roi;
            
    [abs_dist_nearest_roi, dist_nearest_roi, index_dist_nearest_roi]= getNearestRegionOfInterest(CvpXRadians)
    

    // if the RoI is under 30° from the current center of the viewport, it is not needed to do a snapcut
    if (abs_dist_nearest_roi >= 0.5235){
                    
        fireRotation(dist_nearest_roi);
    }
    else {
        doIt = false;
    }

    return dist_nearest_roi;

}

function getNearestRegionOfInterest(CvpXRadians)
{
    let abs_dist_nearest_roi = Infinity;
    let dist_nearest_roi;
    let index_dist_nearest_roi = 0;

    for (let i in editEditInfoJSON[next_edit]["region_of_interest"])
    {
        let ROIXRadians = convert_normalized_to_radians(editEditInfoJSON[next_edit]["region_of_interest"][i]["ROI_theta"]);
        
        let CvpXOpositeRadians = CvpXRadians > 0 ? CvpXRadians - Math.PI : CvpXRadians + Math.PI;

        let curr_dist_roi = getSphereRotation(CvpXRadians, CvpXOpositeRadians, ROIXRadians);
        
        if (Math.abs(curr_dist_roi) < abs_dist_nearest_roi){
            index_dist_nearest_roi = i;
            abs_dist_nearest_roi = Math.abs(curr_dist_roi);		
            dist_nearest_roi = curr_dist_roi;

        }
    }

    return [abs_dist_nearest_roi, dist_nearest_roi, index_dist_nearest_roi]
}

function getSphereRotation(CvpXRadians, CvpXOpositeRadians, ROIXRadians)
{
    if ((CvpXRadians > 0 && ROIXRadians > 0) || (CvpXRadians < 0 && ROIXRadians < 0))
    {
        return ROIXRadians - CvpXRadians;
    }
    
    if (CvpXOpositeRadians > 0)
    {
        // CvpXRadians is negative and ROIXRadians is positive
        if (CvpXOpositeRadians < ROIXRadians)
        {
            //Rotates to the left
            return -1*((Math.PI + CvpXRadians) + (Math.PI - ROIXRadians));
        }
        else
        {
            //Rotates to the right
            return ROIXRadians - CvpXRadians;
        }
    }
    else
    {
        // CvpXRadians is positive and ROIXRadians is negative
        if (CvpXOpositeRadians < ROIXRadians)
        {
            //Rotates to the left
            return -1*(CvpXRadians - ROIXRadians);
        }
        else
        {
            //Rotates to the right
            return (Math.PI - CvpXRadians) + (Math.PI + ROIXRadians);
        }
        
    }
}

function convert_normalized_to_degree(cvp_norm) {
    return 360*cvp_norm - 180;
};

function convert_normalized_to_radians(cvp_norm) {
    return 2*Math.PI*cvp_norm - Math.PI;
};